### ISSUE TITLE
Recognising character literals

### ISSUE THREAD

**Issue by twesterhout** [twesterhout/splc#5]
- [ ] Basic regex
- [ ] Escapes??

**Comment by Rera93** [comment]
I think the special escaping character should be ignored. What do you think?
Also, do you know which regex library is better to use?

**Comment by Rera93** [comment]
`import Text.Regex.PCRE
data TokenCharLit = Error | CharLitVal [Char] 
scanner :: [Char] -> TokenCharLit
scanner input = s0 input
  where 
    s0 [s : r] 
      | s = '\''    = s1 [s] r 
      | otherwise   = Error
    s1 t [s : r] 
      | s =~ "regex" :: Bool = s2 [s] r 
      | otherwise       = Error 
    s2 t [s : r]  
      | s = '\''    = s2 (t++[s]) r 
      | otherwise   = Error    
    s2 t input = CharLitVal t    `

**Comment by Rera93** [comment]
```
import Text.Regex.PCRE

data TokenCharLit = Error | CharLitVal [Char]
 
scanner :: [Char] -> TokenCharLit

scanner input = s0 input
  where 
    s0 [s : r] 
      | s = '\''    = s1 [s] r 
      | otherwise   = Error
    s1 t [s : r] 
      | s =~ "regex" :: Bool = s2 [s] r 
      | otherwise       = Error 
    s2 t [s : r]  
      | s = '\''    = s2 (t++[s]) r 
      | otherwise   = Error    
    s2 t input = CharLitVal t    
```

**Comment by Rera93** [comment]
This is how I think of the character literal recognizer. 
The pattern of the regex is to be decided I guess. Are we gonna exclude special chars? 

Also, I dont know if this regex lib is the right one to use. 
And sorry for writting the scanner in same clean form as before. You can explain me how its properly done the next time we meet. Important is for you to get the idea

**Comment by twesterhout** [comment]
1) Why would you need a regex library at all?
2) Why would you need a list of `Char` in your data type?
3) I don't understand the `s =~ "regex"` part. `s` is a `Char`, so why'd you run a regex match on a single character??
4) Also, your algorithm never terminates unless it encounters an error...
5) Just use Haskell's syntax rather than Clean's ;)

**Comment by Rera93** [comment]
Well in order to match any Char, I thought of using some regex to recognize 'any' char literal? Can it be done differently? 
I thought of a list since it need to recognize [ '\'', 'anyChar', '\''].   
Yeah you are right about the single character thing.
I need to test it with the ghci but I need to get the syntax right in Haskell first :/.

**Comment by twesterhout** [comment]
Yes, have a look at [Data.Char](https://hackage.haskell.org/package/base-4.10.1.0/docs/Data-Char.html) module.

Ah, I get your idea. But do you actually need to store the quotes?

Well, what's the problem? Or are you saying you are not familiar with Haskell's syntax?
